More intuitive history heuristics

aartaka opened the pull request:
# Description

This adds two history improvements utilizing the `htree` and Nyxt features:
- Remembering the requests that were cancelled mid-air, and adding those to history as parents of the ones that replace them.
- Going back in history instead of creating a new child, if the URL opened is already present on the current branch. This way, we utilize the branching nature of the history and re-create the structure of the page in the history tree. The current implementation has downsides:
  - If the "hub" URL (i.e., the one that person returns to) is not the first one they visit, then some of the URLs visited from the hub might actually be recognized as parents of the hub. We need to intelligently sort the history for parent-child/index-page relations and swap the neighbouring nodes if we see them being thus related and misplaced. This all might've sounded unintelligeble, so here's an illustration of what the current logic does and how it might fail:
```
github.com/.../pull/1 ->  github.com/notifications -> github.com/.../pull/3 -> github.com/notifications
                                                                               ^ current node here

Gets optimized into this structure with this PR:

github.com/.../pull/1 ->  github.com/notifications -> github.com/.../pull/3
                          ^ current node here

But then, if one decides to visit github.com/.../pull/1 from the "hub" (github.com/notifications), 
then we end up with a mental mismatch of a model with history structure:

github.com/.../pull/1 ->  github.com/notifications -> github.com/.../pull/3
^ current node here

While it should've been
github.com/notifications -> github.com/.../pull/3
                       |-> github.com/.../pull/1
                           ^ current node here

Because the mental map of the user most probably has /notifications 
as the hub URL that all the journeys start from. github.com/.../pull/1 
is merely a coincidental URL that the browsing started from in one particular session.
```

# Discussion

I'll be happy if we resolve some more problems of our history interface (@jmercouris, note that I'm saying interface here, it's not about `htree` as a structure, it's about us not watching all the corner-cases and blindly putting things into `htree`) as part of this PR, but I am okay with posponing it until later:
- Better canceled link inspection. Simply putting them into history is not always the right decision (can't remember the cases that break it, but there certainly were some, because of which I didn't open this PR three months ago). We need better heuristics for when link canceling is meaningful, and when it's not.
- Properly handling redirection URLs. Right now if there'a redirect from `foo.org#hello` to `foo.org#hi`, we only remember the first one and ignore the second one based on it having the "same URL". There are two issues at play here:
  - Flawed (in some cases like this) URL equality operators, easy to fix.
  - Broken redirection storage, only storing the first URL that started redirect chain. I am not certain about the reasons for that. 
  - I guess that the right behavior for redirects would be to store only the final redirect chain URL in history.
    - Alternatively, we can store all the redirections and all the chain links in history, but mark them as such, so that when one goes `history-backwards` to a redirect URL, they get prompted or automatically intuitively moved to before it, to not trigger the inifinite cycle of redirections.

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.
- [X] I have pulled from master before submitting this PR.
- [X] There are no merge conflicts.
- [X] I've added the new dependencies as:
  - [X] ASDF dependencies,
  - [X] Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - [X] and Guix dependencies.
- [X] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [X] I have performed a self-review of my own code.
- [x] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [x] Documentation:
  - [x] All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - [x] I have updated the existing documentation to match my changes.
  - [X] I have commented my code in hard-to-understand areas.
  - [x] I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - [x] I have added a `migration.lisp` entry for all compatibility-breaking changes.
- [X] Compilation and tests:
  - [X] My changes generate no new warnings.
  - [x] I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - [X] New and existing unit tests pass locally with my changes.

aadcg commented:
I'm stepping out of this discussion as I don't know how our history facilities evolved over times.

aartaka commented:
Well, you can compare the current state of things with this PR and report whether it's convenient or not, at least (: This PR, after all, is not about structure, but rather about the UX.

Ambrevar commented:
The first cancellation fix is welcome, of course!

About hub-URL-revisitation, since it's debatable but has its uses, I suggest to make it an option.
We could make add a slot to the `history` class, but that would mean toggling the option globally, + it's yet another change to this library.

Can't we implement the same thing from `mode/history.lisp`?  Should be easy enough: check if a parent matches, if so call `htree:visit-all`, if not call `htree:add-child`.

Then this behaviour can become a mode option for history-mode, and thus be local.  Thoughts?

aartaka commented:
> About hub-URL-revisitation, since it's debatable but has its uses, I suggest to make it an option. We could make add a slot to the `history` class, but that would mean toggling the option globally, + it's yet another change to this library.
> 
> Can't we implement the same thing from `mode/history.lisp`? Should be easy enough: check if a parent matches, if so call `htree:visit-all`, if not call `htree:add-child`.

Indeed, sounds like a better option to me!

aartaka commented:
@Ambrevar, redirections problem reported in #2660.

aartaka commented:
Hub back-tracking implemented in (failing) d8cac127efa2ad26759ed2a665a0931f74f05058.

Ambrevar commented on source/mode/history.lisp:
```diff
@@ -21,6 +21,11 @@ search.")
+   (backtrack-to-hubs-p
+    nil
+    :type boolean
+    :documentation "Go back in history when encountering an already visited page.
```
Maybe add an example here, because this sentence alone is not very clear.

Ambrevar commented on source/mode/history.lisp:
```diff
@@ -381,6 +386,32 @@ internally, but this display is clearer and more navigable."
+                            :default (nyxt::make-history-tree))
+    (unless (or (url-empty-p url)
+                ;; If buffer was not registered in the global history, don't
+                ;; proceed.  See `buffer's `customize-instance' `:after' method..
```
Spurious double "..".

Ambrevar reviewed (commented):
Looks good to me!

aartaka commented:
I've added some tests for history backtracking, and they seem to fail, although backtracking works just fine when using Nyxt from this branch. @Ambrevar, any idea which part of the state I may be missing here?

Ambrevar commented:
The error is 

```lisp
  | ERROR: There is no applicable method for the generic function
  |          #<STANDARD-GENERIC-FUNCTION HISTORY-TREE:DATA (2)>
  |        when called with arguments
  |          (NIL).
```

You'd need a backtrace to be sure (you can run lisp-unit2 tests with a debugging context), but I'm guessing the error comes from.

I don't think the problem is with state, but rather with timing: the programmatic version might be too fast and thus some history operations might get skipped.

aartaka commented:
That is strange. Should it not synchronize the contents of the file between calls to `files:content`?

Ambrevar commented:
> Should it not synchronize the contents of the file between calls to files:content?

Not that simple: it does, but writing to disk might be slower than the call, so if you check the content of the file externally, it might not reflect the content in memory.

aartaka commented:
> > Should it not synchronize the contents of the file between calls to files:content?
> 
> Not that simple: it does, but writing to disk might be slower than the call, so if you check the content of the file externally, it might not reflect the content in memory.

Maybe block `files:content` until the disk operations are done then?

Ambrevar commented:
No, because the point of `nfiles:content` is to be fast. 

Actually `nfiles:content` has a `wait-p` option.  So this guarantees that the threaded operation is done upon return whant `wait-p` is T.  However this does not guarantee that the OS is done reading/writing the file.

Ambrevar commented:
In any case, I don't have a clear idea of the problem.  Not sure it has to do with Nfiles.  I'd need to have a closer look.

aadcg commented:
What's the status here?

aartaka commented:
Need to rebase and test this with the updates to `nfiles` and see whether those fix the problem of `nfiles:content` returning outdated results sometimes.

aartaka commented:
Rebased over `master`, squashed all the fix/patch commits, ensured all the experimental features (hub backtracking) are behind an opt-in flag—should be ready to merge!

aartaka commented:
The problem I attributed to Nfiles was rather the problem of parent-finding logic in my code （＠´＿｀＠）

Should be fixed now.

aartaka commented:
I just realized we have no manual entry for history and its configuration. Have to fix it. 

@aadcg, if you feel like another round of review, then I'll be glad to see your comments. In the meanwhile, I'll write up the history entry of the manual!

aadcg commented on source/manual.lisp:
```diff
@@ -243,6 +243,45 @@ follows.")
 
+      (:nsection :title "History"
+        (:p "Nyxt history model is a tree—and it's a tree that branches out through all the
+buffers. If you create a new buffer (via " (:nxref :command 'nyxt/hint-mode:follow-hint-new-buffer)
```
I think the first sentence should read: "Nyxt history model is a tree whose nodes are URLs."

aadcg commented on source/manual.lisp:
```diff
@@ -243,6 +243,45 @@ follows.")
+        (:p "Nyxt history model is a tree—and it's a tree that branches out through all the
+buffers. If you create a new buffer (via " (:nxref :command 'nyxt/hint-mode:follow-hint-new-buffer)
+" or " (:nxref :command 'make-buffer) "), it becomes a new history branch originating
+from the branch of the current buffer.")
```
It's not very explicit for the reader. The concept buffer is very confusing because it's never clear whether it's the one before or after running a command that creates a buffer and focus it.

aadcg commented on source/manual.lisp:
```diff
@@ -243,6 +243,45 @@ follows.")
+Nyxt 2, where the history was still a tree, but was not spilling across the
+buffers, then configure "
+            (:nxref :slot 'nyxt/history-mode:conservative-history-movement-p
+              :class-name 'nyxt/history-mode:history-mode)
```
This a doubt I have as a user. What's the difference between `conservative-history-movement-p` and `global-history-p`?

aartaka replied:
Waaaaaaaait...

aadcg commented on source/manual.lisp:
```diff
@@ -243,6 +243,45 @@ follows.")
+            (:nxref :command 'nyxt/history-mode:history-forwards)
+            ") will only work inside the buffer history then.")
+        (:p "Nyxt supra-buffer history has benefits, though: it optimizes browsing patterns
+into something more or less intuitive and productive. One particular pattern
```
"more or less" is not an expression suitable to be used in a manual.

aartaka replied:
Right!

aadcg commented on source/mode/history.lisp:
```diff
@@ -21,6 +21,20 @@ search.")
+
+github.com/.../pull/1 ->  github.com/notifications -> github.com/.../pull/3
+                          ^ current node here
+
```
I think it would be more informative and concise to name these example URLs differently: "URL-A", "URL-B" and "URL-C".

Ambrevar commented on tests/offline/global-history.lisp:
```diff
@@ -40,10 +43,26 @@
+        (nyxt/history-mode::history-add (quri:uri "http://example.org"))
+        (assert-eq 2
+                   (length (htree:all-data (files:content file))))
+        (sleep 0.2)
```
Why the `sleep` calls?

aartaka replied:
It was an artifact of the times I tried to fix the problem of backtracking not getting tested. Removed now!

Ambrevar commented on source/mode/history.lisp:
```diff
@@ -382,6 +396,31 @@ internally, but this display is clearer and more navigable."
+                ;; proceed.  See `buffer's `customize-instance' `:after' method.
+                (not (htree:owner history (id buffer))))
+      (alex:if-let ((parent-position (and (find-submode 'history-mode)
+                                          (backtrack-to-hubs-p (find-submode 'history-mode))
```
Let-bind `history-mode`.

aartaka replied:
Done! A shame there's no `if-let*` in any of the libraries we use...

Ambrevar reviewed (commented):
Looks good to me beside 2 nits :)

aartaka commented:
Should be all covered now!

Ambrevar commented:
Nice, feel free to merge.

Ambrevar commented:
One last nit: Maybe you can extend the `history-add` docstring to explain the new optional behaviour.

In any case, feel free to merge.

aartaka commented:
Docsting extended, commits squashed, branch merged :D

aadcg commented on source/manual.lisp:
```diff
@@ -243,6 +243,49 @@ follows.")
                                       %slot-value%))))))
 
+      (:nsection :title "History"
+        (:p "Nyxt history model is a tree whose nodes are URLs. It branches out through all
```
Pity that it has been merged already. Typo "Its".

aadcg commented:
The manual section is not satisfactory but I'll rewrite it some day.
